projects/interfaces/buffer_test.go

```diff
@@ -0,0 +1,123 @@
+package interfaces
+
+import (
+	"reflect"
+	"testing"
+)
+
+func TestBuffer(t *testing.T) {
+	input := []byte("Hello world!")
+
+	t.Run("return same bytes as start byte", func(t *testing.T) {
+		var b OurByteBuffer
+		b.Write(input)
+		got := b.Bytes()
+		want := input
+
+		assertCorrectMessage(t, got, want)
+	})
+
+	t.Run("write extra bytes to buffer", func(t *testing.T) {
+		var b OurByteBuffer
+		firstInput := "Hello "
+		secondInput := "world"
+		b.Write([]byte(firstInput))
+		b.Write([]byte(secondInput))
+		got := b.Bytes()
+		want := []byte(firstInput + secondInput)
+		assertCorrectMessage(t, got, want)
+
+	})
+
+	t.Run("read from buffer", func(t *testing.T) {
+		var b OurByteBuffer
+		readbuf := make([]byte, len(input))
+		b.Write(input)
+		bytesRead, err := b.Read(readbuf)
+
+		if err != nil {
+			t.Errorf("expect error to be nil, got %v\n", err)
+		}
+		expectedNum := len(input)
+
+		if bytesRead != expectedNum {
+			t.Errorf("got %v but expected %v\n", bytesRead, expectedNum)
+		}
+		assertCorrectMessage(t, readbuf, input)
+
+	})
+
+	t.Run("read: returns error when buffer is empty", func(t *testing.T) {
+		var b OurByteBuffer
+		readbuf := make([]byte, len(input))
+		bytesRead, err := b.Read(readbuf)
+
+		if err.Error() != "EOF" {
```

Comment:
I'd probably check `errors.Is(err, io.EOF)` rather than checking the stringification, though this does work too. Can you talk through what the difference is?

Reply:
I saw something interesting in the documentation of `errors.Is` - It still detects the target error even when it has been wrapped by fmt.Errorf. If the target error is wrapped, the error string will be equal to what it has been wrapped with, so stringfication won't be ideal.